python/lsst/daf/butler/direct_butler/_direct_butler.py

```diff
@@ -1443,27 +1443,40 @@ def _exists_many(
 
         return existence
 
-    def removeRuns(self, names: Iterable[str], unstore: bool = True) -> None:
+    def removeRuns(
+        self, names: Iterable[str], unstore: bool = True, *, unlink_from_chains: bool = False
+    ) -> None:
         # Docstring inherited.
         if not self.isWriteable():
             raise TypeError("Butler is read-only.")
         names = list(names)
         refs: list[DatasetRef] = []
-        all_dataset_types = [dt.name for dt in self._registry.queryDatasetTypes(...)]
+        # Map of the chained collections to the RUN children.
+        parents_to_children: dict[str, set[str]] = defaultdict(set)
+
         with self._caching_context():
-            for name in names:
-                collectionType = self._registry.getCollectionType(name)
-                if collectionType is not CollectionType.RUN:
-                    raise TypeError(f"The collection type of '{name}' is {collectionType.name}, not RUN.")
-                with self.query() as query:
-                    # Work out the dataset types that are relevant.
-                    collections_info = self.collections.query_info(name, include_summary=True)
-                    filtered_dataset_types = self.collections._filter_dataset_types(
-                        all_dataset_types, collections_info
-                    )
-                    for dt in filtered_dataset_types:
-                        refs.extend(query.datasets(dt, collections=name))
+            # Get information about these RUNs.
+            collections_info = self.collections.query_info(names, include_parents=unlink_from_chains)
+            for info in collections_info:
+                if info.type is not CollectionType.RUN:
+                    raise TypeError(f"The collection type of '{info.name}' is {info.type.name}, not RUN.")
+                if unlink_from_chains:
+                    if info.parents is None:  # For mypy.
+                        raise AssertionError("Internal error: Collection parents required but not received")
+                    for parent in info.parents:
+                        parents_to_children[parent].add(info.name)
+
+            # Get all the datasets from these runs.
+            refs = self._query_all_datasets(
+                [info.name for info in collections_info], find_first=False, limit=None
+            )
+
         with self._datastore.transaction(), self._registry.transaction():
```

Comment:
Question is whether this should do something like:

```python
for chunk in chunk_iterable(refs, chunk_size=500_000):
    with self._datastore.transaction(), self._registry.transaction():
        unlink _from chains
        datastore.trash(chunk)
        pruneDatasets(refs, ...)

    datastore.emptyTrash(chunk)
registry.removeCollection(name)
```
on the basis that if it fails inside a chunk, rerunning the command should work just fine and it shouldn't be a problem if some datasets were deleted -- you wanted them deleted anyhow.